### PR TITLE
SI-8366 make partial function and match trees disjoint

### PR DESCRIPTION
**Pull request by densh**
Previously one could match a partial function with match quasiquote:

```
scala> val q"$scrutinee match { case ..$cases }" = q"{ case Foo => Bar
}"
scrutinee: universe.Tree = <empty>
cases: List[universe.CaseDef] = List(case Foo => Bar)
```

This was quite annoying as it leaked encoding of partial functions as
Match trees with empty tree in place of scrutinee.

This commit make sure that matches and partial functions are disjoint
and don't match one another (while preserving original encoding under
the hood out of sight of the end user.)

review @retronym

### CODE REVIEW

File: test/files/scalacheck/quasiquotes/TermConstructionProps.scala [hunk 1]
```diff
@@ -95,12 +95,6 @@ object TermConstructionProps extends QuasiquoteProperties("term construction") {
     body1 ≈ body && cond1 ≈ cond
   }
 
-  property("unquote trees into alternative") = forAll { (c: Tree, A: Tree, B: Tree) =>
-    q"$c match { case $A | $B => }" ≈
-      Match(c, List(
-        CaseDef(Alternative(List(A, B)), EmptyTree, Literal(Constant(())))))
-  }
-
```
Comment (densh) [inline test/files/scalacheck/quasiquotes/TermConstructionProps.scala:103]: This test was quite outdated. Alternatives are tested in PatternConstructionProps now.

**Comment by xeno-by** [comment]
LGTM looks like a step in the right direction.

File: test/files/scalacheck/quasiquotes/TermConstructionProps.scala [hunk 2]
```diff
@@ -295,4 +289,12 @@ object TermConstructionProps extends QuasiquoteProperties("term construction") {
...
     assert(a ≈ q"a" && b ≈ q"b" && c ≈ q"c" && d ≈ q"d" && e ≈ q"e" && g ≈ q"g" && h ≈ q"h" && k ≈ q"k" && l ≈ q"l")
   }
+
+  property("match scrutinee may not be empty") = test {
+    assertThrows[IllegalArgumentException] {
+      val scrutinee = q""
+      val cases = List(cq"_ =>")
+      q"$scrutinee match { case ..$cases }"
+    }
+  }
```
Comment (xeno-by) [inline test/files/scalacheck/quasiquotes/TermConstructionProps.scala:307]: Do we also have tests for construction and deconstruction of partial functions?

Comment (densh) [inline test/files/scalacheck/quasiquotes/TermConstructionProps.scala:307]: Good point, will add some more.

Comment (densh) [inline test/files/scalacheck/quasiquotes/TermConstructionProps.scala:307]: Fixed.

**Comment by adriaanm** [comment]
This will need a rebase

**Comment by densh** [comment]
PLS REBUILD/pr-scala@f6712e4

**Pull request merged**